lib/src/command/add.dart

```diff
@@ -170,42 +237,73 @@ class AddCommand extends PubCommand {
   /// If a version constraint is provided when the `--path` or any of the
   /// `--git-<option>` options are used, a [PackageParseError] will be thrown.
   ///
-  /// If both `--path` and any of the `--git-<option>` options are defined,
-  /// a [PackageParseError] will be thrown.
+  /// Packages must either be a git, hosted, sdk, or path package. Mixing of
+  /// options is not allowed and will cause a [PackageParseError] to be thrown.
   ///
   /// If any of the other git options are defined when `--git-url` is not
   /// defined, an error will be thrown.
   Pair<PackageRange, dynamic> _parsePackage(String package) {
     ArgumentError.checkNotNull(package, 'package');
 
-    if ([path != null, hasGitOptions, hasHostOptions, sdk != null]
-            .where((element) => element == true)
-            .length >=
-        2) {
-      usageException(
-          'Packages must either be a git, hosted, sdk, or path package.');
+    final _conflictingFlagSets = [
+      ['git-url', 'git-ref', 'git-path'],
+      ['hosted-url'],
+      ['path'],
+      ['sdk'],
+    ];
+
+    for (final flag
+        in _conflictingFlagSets.expand((s) => s).where(argResults.wasParsed)) {
+      final conflictingFlag = _conflictingFlagSets
+          .where((s) => !s.contains(flag))
+          .expand((s) => s)
+          .firstWhere(argResults.wasParsed, orElse: () => null);
+      if (conflictingFlag != null) {
+        usageException(
+            'Packages can only have one source, pub add flags "--$flag" and '
+            '"--$conflictingFlag" are conflicting.');
+      }
     }
 
+    /// The package to be added, along with the user-defined package constraints
+    /// if present.
     PackageRange packageRange;
+
+    /// The entry to be added to the pubspec. Assigned dynamic because it can
+    /// take on either a string for simple version constraints or a map for
+    /// more complicated hosted/git options.
     dynamic pubspecInformation;
 
     final splitPackage = package.split(':');
     final packageName = splitPackage[0];
 
     /// There shouldn't be more than one `:` in the package information
     if (splitPackage.length > 2) {
-      throw FormatException('Invalid package and version constraint: $package');
+      usageException('Invalid package and version constraint: $package');
     }
 
     /// We want to allow for [constraint] to take on a `null` value here to
     /// preserve the fact that the user did not specify a constraint.
-    final constraint = splitPackage.length == 2
-        ? VersionConstraint.parse(splitPackage[1])
-        : null;
+    VersionConstraint constraint;
+
+    try {
+      constraint = splitPackage.length == 2
+          ? VersionConstraint.parse(splitPackage[1])
+          : null;
+    } on FormatException catch (e) {
+      usageException('${e.message} Ensure that your version constraint '
+          'satisfies pub semver requirements.');
```

Comment:
Can't we just say something like: `Invalid version constraint: ${e.message}`

Do we have a test case for this? I don't see any containing this message.

Reply:
Modified the message!

There was a test `test/common/add_test.dart` at lines 263 - 284 but it tested for `${e.message}` rather than the full error message. Fixed it!